Documentation/Relationship/Relationship.md

```diff
@@ -0,0 +1,333 @@
+
+The path structure of the Webservice can be used to infer relationship information.  
+For example the fact that the `/user/:userid/post` path is a subroute to `/user/:userid` path can be used to 
+generate linking information for every user, namely a link pointing to `user/{userid}/post`.
+
+In general every endpoint will link to **all** of its children/subroutes.
+
+Giving an example: When a client sends a request for `/user/532` the a potential REST exporter will return the following:
+```json
+{
+  "id": "532",
+  "name": "Rudi",
+  "lastname": "Raser",
+  "email": "rudi@raser.de",
+  "_links": {
+    "self": "https://example.api/user/532",
+    "post": "https://example.api/post"
+  }
+}
+```
+
+One of the core concepts of REST (with HATEOAS) is that a client does not need to know anything about the path structure.
+Instead, just like a web browser, a client follows hyperlinks returned from the web service.  
+This allows you to change the path structure of your REST service - keeping the same named links at the same places -
+without breaking compatibility for your clients.
+
+To support such scenarios there must be some sort of way to manually overwrite such relationship information.
+
+## 2. Manually defining "structural" relationship information
+
+This chapter imagines ways how a user could manually override or provide additional relationship information.
+
+### 2.1. Overriding the link name
+
+The above described inferring approach will use the string description of the `PathComponent` as the link name.
+This may not fit everybody's needs.
+
+Similar to how you can manually define a `@Parameter` name, you can also override a link name.
+This proposal suggest a sort of `PathComponentModifier` (similar to `ComponentModifiers`).
+
+Such a definition could then look like the following:
+```swift
+var content: some Component {
+  Group("test".relationshipName("linkName")) {
+    Handler()
+  }
+}
+```
+
+### 2.2. Adding new links
+
+In order to support rearranging components, a user can also define its own relationships.
+We introduce a `.relationship` modifier for `PathComponent`s, to which one would pass
+a relationship name and `PathComponent`s describing the destination.  
+The destination can of course also include `@PathParameters`.
+
+Any manual definition of a relationship MUST always override a inferred relationship if the name collides.
+
+Looking at the example below, a REST exporter would add the link `greeter` pointing to `/greeting/{userid}` on a 
+response returned from `/user/:userId`. As the value of the `userId` `PathParameter` is known it is replaced with
+the actual value. `@PathParameter`s for which no predefined value can be inferred should be formatted according to
+[RFC 6570](https://tools.ietf.org/html/rfc6570) (as well as potential query parameters of the destination). 
+
+```swift
+var content: some Component {
+  Group("user", $userId.relationship("greeter", to: "greeting", $userId)) {
```

Comment:
I'm not sure about this specific scenario. 
We don't have a way to enforce at compile time, that this path exists.

If I understand this correctly you want to create a link:
```
"_links": {
    "greeting" : "/greeting/12345"
}
```

And while I think this is a cool and valid use case, I'm not sure if we need to address this right away. 

PS: I think we could derive it by the use of the shared parameter `userId`. And not require any explicit path...

Reply:
> I think we could derive it by the use of the shared parameter userId

The use of the `userId` parameter in this example was really only to demonstrate that this mechanism could also perfectly handle path parameters. Yes, in this particular example one could probably somehow create the same result by looking at locations where the same parameter is used multiple times, but the general problem of defining manual additions to sort of "random" locations isn't solved with that.

I agree that the need of providing the full path isn't a very elegant solution, I couldn't really think about a better solution.

This mechanism is really only meant to solve the problem described in [1.](https://github.com/Apodini/Apodini/blob/feature/realtionship/Documentation/Relationship/Relationship.md#1-using-the-path-structure): 
> One of the core concepts of REST (with HATEOAS) is that a client does not need to know anything about the path structure. Instead, just like a web browser, a client follows hyperlinks returned from the web service.
This allows you to change the path structure of your REST service - keeping the same named links at the same places - without breaking compatibility for your clients.

If I may give you another (more abstract) example:

Transforming ...
```swift
var content: some Component {
  Group("a") {
    AHandler()
    Group("b") {
      BHandler()
    } 
  }
}
```
... to ...
```swift
var content: some Component {
  Group("a") {
    AHandler()
  }
  Group("b") {
    BHandler()
  } 
}
```

should be possible without breaking (REST) client compatibility when keeping a link from `a` -> `b`.
This mechanism would allow for that.

But I see your point. We could also just agree on not allowing manual linking at all and force users to do API versioning instead. Or maybe we envision a better idea in the future which solves all this.

Reply:
Throwing in a argument against this part of my proposal: 

Regarding the point
> Any manual definition of a relationship MUST always override an inferred relationship if the name collides.

This works for REST, but for GraphQL this would essentially render the whole sub route, which is overwritten by the manual definition, inaccessible. 🤔 So at least that part must be changed, so that Apodini must not allow overriding existing relationship definitions (or name collisions with sub components).

Reply:
See https://github.com/Apodini/Apodini/pull/41#discussion_r541122416 (moving discussion to that subthread)

> The revised proposal removed the need of specifying that redundant information. Instead I introduced a approach similar to what you suggested [2.2](https://github.com/Apodini/Apodini/blob/feature/realtionship/Documentation/Relationship/Relationship.md#22-adding-new-relationships).
> 
> Would love to hear your feedback on this, as I think specifying `KeyPaths` with a global/centralized Relationship definition is pretty elegant, but not sure if I delivered the best possible solution.